packages/proxy/lib/http/util/cookies.ts

```diff
@@ -0,0 +1,166 @@
+import _ from 'lodash'
+import type Debug from 'debug'
+import { URL } from 'url'
+import { cors } from '@packages/network'
+import { Cookie, CookieJar } from '@packages/server/lib/cookie-jar'
+
+interface AutomationCookie {
+  domain: string
+  expiry: number | null
+  httpOnly: boolean
+  maxAge: number | null
+  name: string
+  path: string | null
+  sameSite: string
+  secure: boolean
+  value: string
+}
+
+interface RequestDetails {
+  url: string
+  isAUTFrame: boolean
+  needsCrossOriginHandling: boolean
+}
+
+// Sets SameSite context to match what it would be in the browser
+// see https://github.com/salesforce/tough-cookie#samesite-cookies
+export const getSameSiteContext = (autUrl: string | undefined, requestUrl: string, isAUTFrameRequest: boolean) => {
+  // if there's no AUT URL, it's a request for the first URL visited,
+  // so there would be no cross-origin concerns
+  // if the request origin matches the AUT origin, cookies can be handled
+  // in a strict fashion
+  if (!autUrl || cors.urlOriginsMatch(autUrl, requestUrl)) {
+    return 'strict'
```

Comment:
How does this work if I say have a site `foobar.com` and a cookie that I set `SameSite=none` in a first party context? If `foobar.com` sends requests to the `foobar.com` domain, they will be set as `SameSite=strict` cookies, but if I say do a cross origin navigation to `barbaz.com`, and send a request to `foobar.com`, would the cookie then be `SameSite=none`? Hopefully that made some sort of sense 😅

Reply:
This property (sameSiteContext) gets tough-cookie on the same page as the browser, so it can simulate what the browser would do, not to be confused with the SameSite property of the cookie itself. Once tough-cookie has that context, it can evaluate whether a cookie should be set or retrieved based on the sameSiteContext, the request url, and the domain of the cookie.

This is the way I mentally translate the sameSiteContext:
- strict: first-party, top-level navigation
- lax: cross-origin, top-level navigation
- none: any navigation or nested resources

Reply:
I think I am still a bit confused by this. Based on the code it looks like the scenario I mentioned would set the cookie context to `strict` instead of `none`. I think tests here might go a long way to show those unfamiliar with the code how the cookie jar is supposed to function in some of these use cases which would also serve as a documentation artifact.

Reply:
It's confusing because tough-cookie is using the same values as SameSite for this sameSiteContext property, even though they're not the same thing. Not sure I can explain it any better than the [tough-cookie doc](https://github.com/salesforce/tough-cookie#samesite-cookies) I linked to above this function. I'm not sure tests would help since they'd just be testing the logic, which is pretty straightforward from the code itself. I'll update to the comments to hopefully make things a little clearer.

Reply:
Updated the comments in [`4903cb1` (#22320)](https://github.com/cypress-io/cypress/pull/22320/commits/4903cb11b3012b3e9cd3ccff1a807d4615643a3f)

Reply:
OK I think this is finally clicking. The goal of SameSiteContext is to understand the context of which a request is made to see if a cookie applies, so if I make a request on `foobar.com` and I have cookies

```
key | value | samesite | domain

foo    bar      none     foobar.com  
baz    quux     strict   foobar.com
``` 

and the request is to `foobar.com`, the `SameSiteContext` is `strict`, but both cookies are sent, because that is what the browser would do. Hopefully I am getting it 😅 .

Reply:
Yeah, I think you've got it. Took me a while to grok it as well.